xbmc/platform/win32/filesystem/Win32WSDiscovery.cpp

```diff
@@ -0,0 +1,146 @@
+/*
+ *  Copyright (C) 2005-2021 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "Win32WSDiscovery.h"
+
+
+HRESULT CreateClientNotificationSink(CClientNotificationSink** sink)
+{
+  HRESULT hr = S_OK;
+  CClientNotificationSink* tempSink = nullptr;
+
+  if (!sink)
+    hr = E_POINTER;
+  else
+    *sink = nullptr;
+
+  if (hr == S_OK)
+  {
+    tempSink = new CClientNotificationSink();
+
+    if (!tempSink)
+      hr = E_OUTOFMEMORY;
+  }
+
+  if (hr == S_OK)
+  {
+    *sink = tempSink;
+    tempSink = nullptr;
+  }
+
+  if (tempSink)
+  {
+    tempSink->Release();
+    tempSink = nullptr;
+  }
+
+  return hr;
+}
```

Comment:
Could this block be cleaned up to something like the following?

```
HRESULT CreateClientNotificationSink(CClientNotificationSink** sink)
{
  CClientNotificationSink* tempSink = nullptr;

  if (!sink)
    return E_POINTER;
  else
    *sink = nullptr;

  tempSink = new CClientNotificationSink();

  if (!tempSink)
    return E_OUTOFMEMORY;

  *sink = tempSink;
  tempSink = nullptr;

  tempSink->Release();
  tempSink = nullptr;

  return S_OK;
}
```

Reply:
maybe even drop the *sink = nullptr, and just do the *sink = tempsink after te tempsink has been checked/created?

Reply:
Sure. I think it can be simplified a bit further :-)

```c++
HRESULT CreateClientNotificationSink(CClientNotificationSink** sink)
{
  CClientNotificationSink* tempSink = nullptr;

  if (!sink)
    return E_POINTER;

  tempSink = new CClientNotificationSink();

  if (!tempSink)
    return E_OUTOFMEMORY;

  *sink = tempSink;
  tempSink = nullptr;

  return S_OK;
}
```